ruma-client-api/src/r0/typing/create_typing_event.rs

```diff
@@ -24,21 +25,64 @@ ruma_api! {
         #[ruma_api(path)]
         pub room_id: RoomId,
 
-        // TODO: Group the following two body fields into an enum
-
-        /// Whether the user is typing or not. If `false`, the `timeout` key can be omitted.
-        pub typing: bool,
-
-        /// The length of time in milliseconds to mark this user as typing.
-        #[serde(
-            with = "ruma_serde::duration::opt_ms",
-            default,
-            skip_serializing_if = "Option::is_none",
-        )]
-        pub timeout: Option<Duration>,
+        /// Whether the user is typing within a length of time or not.
+        #[serde(flatten)]
+        pub state: Typing,
     }
 
     response: {}
 
     error: crate::Error
 }
+
+/// A mark for whether the user is typing within a length of time or not.
+#[derive(Clone, Copy, Debug)]
+pub enum Typing {
+    /// Not typing.
+    Stop,
+    /// Typing during the specified length of time.
+    During(Duration),
+}
+
+#[derive(Deserialize)]
+struct TypingInner {
```

Comment:
If we do this, we might as well derive `Deserialize` and `Serialize` for the public struct and use `serde(from)` + `serde(into)`.

Reply:
Didn't know about those serde attributes, will use them. Also I thought initially that there was one extra unnecessary value instantiation by doing the intermediate conversion step when serializing, but I find my assumption to be wrong.

Reply:
In reading the docs for those two attributes I see that `#[serde(into)]` makes a clone that can be avoided by implementing `Serialize` directly and I can't use `#[serde(from)]` but `#[serde(try_from)]`, and I lose the ability to call the "right" method in serde to display a message error that is consistent with other similiar errors in serde. Should I proceed anyways?

Reply:
I think the clone is easily optimized out and the `serde(into)` version should be slightly simpler.

> I lose the ability to call the "right" method in serde to display a message error that is consistent with other similiar errors in serde.

In that case, let's keep the `Deserialize` impl as-is. It's very short already :)